core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala

```diff
@@ -139,6 +199,356 @@ protected[actions] trait PrimitiveActions {
+   *         or an error
+   */
+  private def invokeComposition(
+    user: Identity,
+    action: ExecutableWhiskActionMetaData,
+    payload: Option[JsObject],
+    waitForResponse: Option[FiniteDuration],
+    cause: Option[ActivationId],
+    caller: Option[Session] = None)(implicit transid: TransactionId): Future[Either[ActivationId, WhiskActivation]] = {
+
+    val session = Session(
+      activationId = activationIdFactory.make(),
+      start = Instant.now(Clock.systemUTC()),
+      action,
+      cause,
+      duration = 0,
+      maxMemory = action.limits.memory.megabytes,
+      state = None,
+      accounting = caller.map { _.accounting }.getOrElse(CompositionAccounting()), // share accounting with caller
+      logs = Buffer.empty,
+      caller,
+      result = waitForResponse.map { _ =>
+        Promise[Either[ActivationId, WhiskActivation]]() // placeholder for result if blocking invoke
+      })
+
+    invokeConductor(user, payload, session)
+
+    // is caller waiting for the result of the activation?
+    waitForResponse
+      .map { timeout =>
+        // handle timeout
+        session.result.head.future
+          .withAlternativeAfterTimeout(timeout, Future.successful(Left(session.activationId)))
+      }
+      .getOrElse {
+        // no, return the session id
+        Future.successful(Left(session.activationId))
+      }
+  }
+
+  /**
+   * A method that knows how to handle a conductor invocation.
+   *
+   * This method prepares the payload and invokes the conductor action.
+   * It parses the result and extracts the name of the next component action if any.
+   * It either invokes the desired component action or completes the composition invocation.
+   * It also checks the invocation counts against the limits.
+   *
+   * @param user the identity invoking the action
+   * @param payload the dynamic arguments for the activation
+   * @param session the session object for this composition
+   * @param transid a transaction id for logging
+   */
+  private def invokeConductor(user: Identity, payload: Option[JsObject], session: Session)(
+    implicit transid: TransactionId) = {
+
+    if (session.accounting.conductors > 2 * actionSequenceLimit) {
+      // composition is too long
+      val response = ActivationResponse.applicationError(compositionIsTooLong)
+      completeAppActivation(user, session, response)
+    } else {
+      // inject state into payload if any
+      val params = session.state
+        .map { state =>
+          Some(JsObject(payload.getOrElse(JsObject()).fields ++ state.fields))
+        }
+        .getOrElse(payload)
+
+      // invoke conductor action
+      session.accounting.conductors += 1
+      val response =
+        invokeSimpleAction(
+          user,
+          session.action,
+          params,
+          Some(session.action.limits.timeout.duration + 1.minute), // wait for result
```

Comment:
i'd use 
```
..., 
waitForResponse = ...,
cause = ...)
```
in addition to/vs the comment.

Reply:
OK